File: backend/src/test/java/dev/coldhands/pair/stairs/backend/ApplicationIT.java
package dev.coldhands.pair.stairs.backend;

import com.jayway.jsonpath.DocumentContext;
import com.jayway.jsonpath.JsonPath;
import dev.coldhands.pair.stairs.backend.domain.*;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.test.context.jdbc.Sql;

import java.time.LocalDate;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class ApplicationIT implements WithBackendHttpClient {

    @Test
    @Sql(value = "/delete-test-data.sql", executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
    @Sql(value = "/delete-test-data.sql", executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
    void basicFlowStartingFromScratch() throws Exception {
        final long dev0Id = createDeveloper("dev-0");
        createDeveloper("dev-1");
        createDeveloper("dev-2");
        createDeveloper("dev-3");

        List<Long> developersToIncludeInCombinations = getDeveloperIdsFor(List.of(
                "dev-0",
                "dev-1",
                "dev-2"
        ));

        createStream("stream-a");
        createStream("stream-b");

        List<Long> streamsToIncludeInCombinations = getStreamIdsFor(List.of(
                "stream-a",
                "stream-b"
        ));

        final List<ScoredCombination> scoredCombinations = calculateCombinations(developersToIncludeInCombinations, streamsToIncludeInCombinations);

        assertThat(scoredCombinations).isNotEmpty();

        final ScoredCombination bestCombination = scoredCombinations.getFirst();

        assertThat(bestCombination.combination().stream()
                .flatMap(ps -> ps.developers().stream())
                .map(DeveloperInfo::displayName)
                .toList())
                .containsExactlyInAnyOrder("dev-0", "dev-1", "dev-2");

        assertThat(bestCombination.combination().stream()
                .map(PairStream::stream)
                .map(StreamInfo::displayName)
                .toList())
                .containsExactlyInAnyOrder("stream-a", "stream-b");


        final LocalDate today = LocalDate.of(2024, 4, 27);
        saveCombinationEventFor(today, bestCombination.combination());

        final List<CombinationEvent> combinationEvents = getCombinationEvents();

        assertThat(combinationEvents).hasSize(1);

        final CombinationEvent savedEvent = combinationEvents.getFirst();

        assertThat(savedEvent.date()).isEqualTo(today);
        assertThat(savedEvent.combination()).hasSize(2);
        assertThat(savedEvent.combination()).isEqualTo(bestCombination.combination());

        deleteCombinationEvent(savedEvent.id());

        final List<CombinationEvent> combinationEventsAfterDelete = getCombinationEvents();
        assertThat(combinationEventsAfterDelete).isEmpty();

        final DeveloperStats developerStats = getDeveloperStatsBetween(dev0Id, today, today);

        assertThat(developerStats.developerStats()).hasSize(4);
        assertThat(developerStats.streamStats()).hasSize(2);
    }

    @Nested
    class Actuator {

        @Test
        void health() {
            final ResponseEntity<String> response = REST_TEMPLATE.getForEntity(BASE_URL + "/actuator/health", String.class);
            assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);

            final String responseBody = response.getBody();
            final DocumentContext parsed = JsonPath.parse(responseBody);

            assertThat(parsed.read("$.status", String.class)).isEqualTo("UP");
            assertThat(parsed.read("$.components.db.status", String.class)).isEqualTo("UP");
            assertThat(parsed.read("$.components.db.details.database", String.class)).isEqualTo("MySQL");
            assertThat(parsed.read("$.components.livenessState.status", String.class)).isEqualTo("UP");
            assertThat(parsed.read("$.components.readinessState.status", String.class)).isEqualTo("UP");
        }

        @Test
        void info() {
            final ResponseEntity<String> response = REST_TEMPLATE.getForEntity(BASE_URL + "/actuator/info", String.class);
            assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);

            final String responseBody = response.getBody();
            final DocumentContext parsed = JsonPath.parse(responseBody);

            assertThat(parsed.read("$.build.version", String.class)).isNotNull();

            assertThat(parsed.read("$.git.branch", String.class)).isNotNull();
            assertThat(parsed.read("$.git.commit.id", String.class)).isNotNull();
        }
    }
}

File: backend/src/test/kotlin/dev/coldhands/pair/stairs/backend/ApplicationIT.kt
package dev.coldhands.pair.stairs.backend

import com.jayway.jsonpath.DocumentContext
import com.jayway.jsonpath.JsonPath
import dev.coldhands.pair.stairs.backend.WithBackendHttpClient.Companion.REST_TEMPLATE
import dev.coldhands.pair.stairs.backend.domain.*
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.collections.shouldContainExactlyInAnyOrder
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.collections.shouldNotBeEmpty
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.test.context.jdbc.Sql
import java.time.LocalDate

class ApplicationIT : WithBackendHttpClient {

    @Test
    @Sql(value = ["/delete-test-data.sql"], executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
    @Sql(value = ["/delete-test-data.sql"], executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
    fun basicFlowStartingFromScratch() {
        val dev0Id = createDeveloper("dev-0")
        createDeveloper("dev-1")
        createDeveloper("dev-2")
        createDeveloper("dev-3")

        val developersToIncludeInCombinations = getDeveloperIdsFor(
            listOf(
                "dev-0",
                "dev-1",
                "dev-2",
            ),
        )

        val streamAId = createStream("stream-a")
        createStream("stream-b")

        val streamsToIncludeInCombinations = getStreamIdsFor(
            listOf(
                "stream-a",
                "stream-b",
            ),
        )

        val scoredCombinations: List<ScoredCombination> =
            calculateCombinations(developersToIncludeInCombinations, streamsToIncludeInCombinations)

        scoredCombinations.shouldNotBeEmpty()

        val bestCombination = scoredCombinations.first()

        bestCombination
            .combination()
            .flatMap { pairStream -> pairStream.developers() }
            .map(DeveloperInfo::displayName)
            .shouldContainExactlyInAnyOrder("dev-0", "dev-1", "dev-2")

        bestCombination
            .combination()
            .map(PairStream::stream)
            .map(StreamInfo::displayName)
            .shouldContainExactlyInAnyOrder("stream-a", "stream-b")

        val today = LocalDate.of(2024, 4, 27)
        saveCombinationEventFor(today, bestCombination.combination())

        val combinationEvents: List<CombinationEvent> = getCombinationEvents()
        combinationEvents shouldHaveSize 1

        val savedEvent = combinationEvents.first()

        savedEvent.date() shouldBe today
        savedEvent.combination() shouldHaveSize 2
        savedEvent.combination() shouldBe bestCombination.combination()

        deleteCombinationEvent(savedEvent.id())

        val combinationEventsAfterDelete: List<CombinationEvent> = getCombinationEvents()
        combinationEventsAfterDelete.shouldBeEmpty()

        val developerStats = getDeveloperStatsBetween(dev0Id, today, today)

        developerStats.developerStats shouldHaveSize 4
        developerStats.streamStats shouldHaveSize 2

        val streamStats = getStreamStatsBetween(streamAId, today, today)

        streamStats.developerStats shouldHaveSize 4
    }

    @Nested
    inner class Actuator {

        @Test
        fun health() {
            val response: ResponseEntity<String> =
                REST_TEMPLATE.getForEntity(
                    WithBackendHttpClient.BASE_URL + "/actuator/health",
                    String::class.java,
                )

            response.statusCode shouldBe HttpStatus.OK

            val parsed: DocumentContext = JsonPath.parse(response.body!!)

            parsed.read("$.status", String::class.java) shouldBe "UP"
            parsed.read("$.components.db.status", String::class.java) shouldBe "UP"
            parsed.read("$.components.db.details.database", String::class.java) shouldBe "MySQL"
            parsed.read("$.components.livenessState.status", String::class.java) shouldBe "UP"
            parsed.read("$.components.readinessState.status", String::class.java) shouldBe "UP"
        }

        @Test
        fun info() {
            val response: ResponseEntity<String> =
                REST_TEMPLATE.getForEntity(
                    WithBackendHttpClient.BASE_URL + "/actuator/info",
                    String::class.java,
                )

            response.statusCode shouldBe HttpStatus.OK

            val parsed: DocumentContext = JsonPath.parse(response.body!!)

            parsed.read("$.build.version", String::class.java).shouldNotBeNull()
            parsed.read("$.git.branch", String::class.java).shouldNotBeNull()
            parsed.read("$.git.commit.id", String::class.java).shouldNotBeNull()
        }
    }
}
